File: frontend/src/models/tileObject.ts
import {CountryIdentifier} from "./country";
import {CityIdentifier} from "./city";

export interface TileObject {
    type: "marker" | "scout" | "city";
    country: CountryIdentifier;
}

export interface MarkerTileObject extends TileObject {
    type: "marker";
    label: string;
}

export interface ScoutTileObject extends TileObject {
    type: "scout";
    creationTurn: number;
}

export interface CityTileObject extends TileObject {
    type: "city";
    city: CityIdentifier;
}

File: backend/strategy-game-common/src/main/kotlin/io/github/smiley4/strategygame/backend/common/models/TileObject.kt
package io.github.smiley4.strategygame.backend.common.models

sealed interface TileObject {
    val countryId: String
}

class MarkerTileObject(
    override val countryId: String,
    val label: String
) : TileObject

class ScoutTileObject(
    override val countryId: String,
    val creationTurn: Int,
) : TileObject

class CityTileObject(
    override val countryId: String,
    val cityId: String,
) : TileObject